### DISCUSSION TITLE
Having issues getting METplotpy line plot to plot values - users_guide example case for line.html

### DISCUSSION THREAD

**Discussion by cpapadop1** [dtcenter/METplus#1594] [Plot Generation] [answered]
Hi everyone,

I am having an issue getting the METplotpy line plot to plot the values from the example case in the tutorial:

https://metplotpy.readthedocs.io/en/latest/Users_Guide/line.html 

along with some other plots, many are working.  I am new to METplotpy/METcalpy so this is probably just me not understanding something.  I have looked through these discussions and was not able to glean an answer.

 Also, what met/metplus tool was used to generate the line.data file in the first place ?

Here are the examples I have tried, the ones that I can not get to work, the points file become N/A for all of them which appears to be the same issue for line plot, box plot, bar plot and maybe ROC diagram:

        line plot                               Data set not quite working with yaml - issue points values
        box plot                               Data set not quite working with yaml - again points
        bar plot                                Data set not quite working with yaml  - again points
        Ensemble spread-skill plot             Works
        contour plots                                  Works
        performance diagram                     Works
        Reliability diagram                          Works  NOTE: does not match online docs but matches png in folder
        ROC diagram                                   Data set not quite working with yaml - again points
        Histogram                                        Works
        2D histogram                                   Works
        Taylor diagram                                 Works

I have tried numerous different conda environments to see if that is it, this does not appear to be the issue.  I am running the default cases from the tutorial, I get N/A in my points files, so that is where the issue is. 

 I am using METplotpy 1.1.0 and METcalcpy 1.1.0

Looking at the line plot to try and debug this:

when I run it I get a bunch of future warnings, these should not be an issue for now, but can't really find any errors:

```
> python $METPLOTPY_BASE/metplotpy/plots/line/line.py $WORKING_DIR/custom_line.yaml
> /home/cpapadop/miniconda3/envs/test_met_7.x/METcalcpy/METcalcpy-1.1.0/metcalcpy/event_equalize.py:140: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
>   discarded_cases = discarded_cases.append(permutation_cases_not_in_common_cases)
> /home/cpapadop/miniconda3/envs/test_met_7.x/METcalcpy/METcalcpy-1.1.0/metcalcpy/event_equalize.py:140: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
>   discarded_cases = discarded_cases.append(permutation_cases_not_in_common_cases)
> /home/cpapadop/miniconda3/envs/test_met_7.x/METcalcpy/METcalcpy-1.1.0/metcalcpy/util/utils.py:710: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
>   all_ee_records = output_ee_data.append(output_ee_data_2).reindex()
> WARNING: eventEqualize() detected non-unique events for ('CONTROL', 'Z02') using [fcst_valid_beg,fcst_lead)]
> WARNING: eventEqualize() detected non-unique events for ('GTS', 'Z02') using [fcst_valid_beg,fcst_lead)]
> /home/cpapadop/miniconda3/envs/test_met_7.x/METcalcpy/METcalcpy-1.1.0/metcalcpy/event_equalize.py:140: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
>   discarded_cases = discarded_cases.append(permutation_cases_not_in_common_cases)
> /home/cpapadop/miniconda3/envs/test_met_7.x/METcalcpy/METplotpy/METplotpy-1.1.0/metplotpy/plots/line/line_series.py:433: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
>   self.series_data = self.series_data.append(stats_indy_1, sort=False)
> /home/cpapadop/miniconda3/envs/test_met_7.x/METcalcpy/METplotpy/METplotpy-1.1.0/metplotpy/plots/line/line_series.py:433: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
>   self.series_data = self.series_data.append(stats_indy_1, sort=False)
> /home/cpapadop/miniconda3/envs/test_met_7.x/METcalcpy/METplotpy/METplotpy-1.1.0/metplotpy/plots/line/line_series.py:433: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
>   self.series_data = self.series_data.append(stats_indy_1, sort=False)
```

The points file is empty:

(test_met_7.x) [cpapadop@comet-07-03 line_plot]$ cat line.points1 
N/A N/A N/A N/A N/A N/A N/A N/A N/A
N/A N/A N/A N/A N/A N/A N/A N/A N/A
N/A N/A N/A N/A N/A N/A N/A N/A N/A

My plot is missing those, but the rest of the plot is complete:

![line](https://user-images.githubusercontent.com/14187732/165407991-b78660ca-c32a-4607-b029-f2eeb5a906ee.png)

Not sure what to do next, I am hoping somebody who is more familiar with these example cases can easily see what is amiss.

Any help suggestions and/or if anybody else was having the same issues and had a solution ??

thanks - Caroline

**Comment by bikegeek** [comment]
Hi Caroline,

Those future and deprecation warnings are not an issue yet, they don't show up for Python 3.6 and their associated third-party Python packages. We are migrating to Python 3.8 and trying to deal with those issues now (github issues created).  

The line plot you created is supposed to use a custom configuration file, custom_line.yaml which is located in the METplotpy/test/line directory.  What command are you using for generating the line plot?  In the default and custom config files, the dump_points settings are set to False, so I'm puzzled as to why you have .points1 files generated.  Can you provide the config file you used?

I'll have to ask Tatiana Burek where she retrieved the data for the line plot.  I'm guessing it came from a METviewer example.  

Summary, I need two things from you:

1) the command you are using to generate the line plot
2) the custom configuration file you are using 

Thanks,
Minna

**Comment by cpapadop1** [comment]
I am using the following command:

> python $METPLOTPY_BASE/metplotpy/plots/line/line.py $WORKING_DIR/custom_line.yaml

The output from that command is in the previous post and I agree the warnings are not an issue until the features are
no longer available.  I am truly puzzled what is not working, it should all work.

I basically made a directory line and copied over:

> cp $METPLOTPY_BASE/test/line/custom_line.yaml $WORKING_DIR/custom_line.yaml .
> cp -pr $METPLOTPY_BASE/test/line/line.data  $WORKING_DIR/line.data

and then updated the custom_line.yaml to change output points to TRUE, and use the local working directory for 
the points output, the original version has both the png file and the input line.data in the local directory which I have, here is a dff from the original file out of the test directory:


(test_met_7.x) [cpapadop@comet-07-03 line_plot]$ diff custom_line.yaml.orig  custom_line.yaml
31c31
< dump_points_1: 'False'
---
> dump_points_1: 'True'
110c110
< #points_path: /dir_to_save_points1_file
---
> points_path: ./ 
218c218
<   type: vert_line
\ No newline at end of file
---
>   type: vert_line


I have tried various different versions of python, but, given dependency issues hard to keep it at 3.6.x:


Here is the full version of the yaml file:

> (test_met_7.x) [cpapadop@comet-07-03 line_plot]$ cat custom_line.yaml
> alpha: 0.05
> box_avg: 'False'
> box_boxwex: 0.2
> box_notch: 'False'
> box_outline: 'True'
> box_pts: 'False'
> caption_align: 0.0
> caption_col: '#333333'
> caption_offset: 3.0
> caption_size: 0.8
> caption_weight: 1
> cex: 1
> colors:
> - '#ff0000'
> - '#55ff00'
> - '#00aaff'
> - '#0040ff'
> - '#c3ff00'
> con_series:
> - 1
> - 1
> - 1
> - 1
> - 1
> create_html: 'False'
> derived_series_1:
> - - CONTROL RH MAE
>   - GTS RH MAE
>   - DIFF
> derived_series_2: []
> dump_points_1: 'True'
> dump_points_2: 'False'
> event_equal: 'True'
> fcst_var_val_1:
>   RH:
>   - MAE
> fcst_var_val_2:
>   TMP:
>   - ME
> fixed_vars_vals_input:
>   fcst_lev:
>     fcst_lev_0:
>     - Z02
> grid_col: '#cccccc'
> grid_lty: 3
> grid_lwd: 1
> grid_on: 'True'
> grid_x: listX
> indy_label:
> - '6'
> - '12'
> - '18'
> indy_stagger_1: 'True'
> indy_stagger_2: 'False'
> indy_vals:
> - '60000'
> - '120000'
> - '180000'
> indy_var: fcst_lead
> legend_box: o
> legend_inset:
>   x: 0.0
>   y: -0.25
> legend_ncol: 3
> legend_size: 0.8
> line_type: None
> list_stat_1:
> - MAE
> list_stat_2:
> - ME
> list_static_val:
>   fcst_var: ME
> mar:
> - 8
> - 4
> - 5
> - 4
> method: bca
> mgp:
> - 1
> - 1
> - 0
> num_iterations: 1
> num_threads: -1
> plot_caption: Caption
> plot_ci:
> - std
> - std
> - std
> - std
> - std
> plot_disp:
> - 'True'
> - 'True'
> - 'True'
> - 'True'
> - 'True'
> plot_filename: ./line.png
> plot_height: 8.5
> plot_res: 72
> plot_stat: median
> plot_type: png16m
> plot_units: in
> plot_width: 11.0
> 
> # Optional, uncomment and set to directory to store the .points1 file
> # that is used by METviewer (created when dump_points_1 is set to True)
> # if dump_points_1 is True and this is uncommented, the points1 file
> # will be saved in the default location (i.e. where the input data file is stored).
> points_path: ./ 
> random_seed: null
> series_line_style:
> - '-'
> - '-'
> - '-'
> - '-'
> - '-'
> series_line_width:
> - 1
> - 1
> - 1
> - 1
> - 1
> series_order:
> - 1
> - 2
> - 3
> - 4
> - 5
> series_symbols:
> - .
> - .
> - .
> - .
> - .
> series_type:
> - b
> - b
> - b
> - b
> - b
> series_val_1:
>   model:
>   - CONTROL
>   - GTS
> series_val_2:
>   model:
>   - CONTROL
>   - GTS
> show_nstats: 'True'
> show_signif:
> - 'False'
> - 'False'
> - 'False'
> - 'False'
> - 'False'
> stat_input: ./line.data
> sync_yaxes: 'False'
> title: test title
> title_align: 0.5
> title_offset: -2
> title_size: 1.4
> title_weight: 2.0
> user_legend: []
> variance_inflation_factor: 'False'
> vert_plot: 'False'
> x2lab_align: 0.5
> x2lab_offset: -0.5
> x2lab_size: 0.8
> x2lab_weight: 1
> x2tlab_horiz: 0.5
> x2tlab_orient: 1
> x2tlab_perp: 1
> x2tlab_size: 0.8
> xaxis: test x_label
> xaxis_reverse: 'False'
> xlab_align: 0.5
> xlab_offset: 2
> xlab_size: 1
> xlab_weight: 1
> xlim: []
> xtlab_decim: 0
> xtlab_horiz: 0.5
> xtlab_orient: 1
> xtlab_perp: -0.75
> xtlab_size: 1
> y2lab_align: 0.5
> y2lab_offset: 1
> y2lab_size: 1
> y2lab_weight: 1
> y2lim: []
> y2tlab_horiz: 0.5
> y2tlab_orient: 1
> y2tlab_perp: 1
> y2tlab_size: 1.0
> yaxis_1: test y_label
> yaxis_2: ''
> ylab_align: 0.5
> ylab_offset: -2
> ylab_size: 1
> ylab_weight: 1
> ylim: []
> ytlab_horiz: 0.5
> ytlab_orient: 1
> ytlab_perp: 0.5
> ytlab_size: 1
> 
> lines:
> - color: '#8000ff'
>   line_width: '2'
>   position: '11'
>   type: horiz_line
>   line_style: '--'
> - line_style: '-'
>   color: '#000000'
>   line_width: 1
>   position: "18"
>   type: vert_line


Please let me know of any other information you need.  

thanks  again,

Caroline

**Comment by bikegeek** [comment, marked as answer]
Thanks Caroline.  I was able to reproduce the NA issue with the line plot.  I'll do some more troubleshooting.  It wasn't anything you had done incorrectly.  Thanks for your patience while we debug this.

**Reply by cpapadop1** [reply]
Wonderful, thank you so much, what MET product produced the line plot data file ??

**Reply by bikegeek** [reply]
I'll find out this afternoon, I have a meeting with the other developers.

**Reply by bikegeek** [reply]
Do you have some MET output data you'd like to try out?  If you can provide it to me I can see if I can get it to work.

**Comment by cpapadop1** [comment]
Just a suggestion, I think the box and bar plots have the same issues as the line plot and it is probably the same issue, the ROC plot has a similar issue and it might be all related.  

As to asking what MET product, I am working with our modelers to get all of this up and running and then moving over to our data, so I don't have a data set per say right now, but, they asked me what MET product produced the output that you used as input for line.data, or was there some post-processing invovled ?

**Reply by bikegeek** [reply]
Hi Caroline, Tatiana Burek is our METviewer developer and the developer for the line plots.  The line.data is data extracted from the METviewer database and not output from a MET tool.  So there isn't any post-processing per se.  

I haven't had a chance to debug the issue yet, but I think most of the plots are using the same general logic for collecting points, so it looks like that's whats happening.  I'll hopefully get to look at things this afternoon.

**Comment by bikegeek** [comment]
I know you cannot use Python 3.6, but I also verified that things are working for Python 3.6.  However, all new development (beta releases, METplotpy v2.0) will be using Python 3.8.  Unfortunately this doesn't help you, but it is an indication that some underlying changes took affect in migrating to a relatively newer version of Python.

**Reply by cpapadop1** [reply]
Ok, I will give it another go to try and get to Python 3.6 again.  Do you know when a release candidate for METplotpy v 2.0 will be available, I could not find "alpha/beta" to be pulled yet, or any hint of a timeline (I will not hold you to it).  If you could give me a conda spec file, I  could try and create the conda environment from that, I just get so many conflicts esp since python 3.6 is now deprecated.  E.g

 conda env export  -n npl > npl-environment.yml

would give me the npl environment from Cheyenne to reproduce.
 
Or I have  Cheyenne accounts and could clone, to try and get a good conda env, referencing

 :https://arc.ucar.edu/knowledge_base/83853599

Thanks for the explanation of the line.data creation, that makes sense to me now.

**Reply by bikegeek** [reply]
We modified our release schedule and we have a coordinated beta1 release (MET, METplus wrappers, METviewer, METdatadb, METcalcpy, and METplotpy) for **June 22**.  That release will have Python 3.8 support.  In the meantime, for METcalcpy, there is a requirements.txt file that you can use to replicate the Python environment: https://github.com/dtcenter/METcalcpy/blob/main_v1.1/requirements.txt. We also have a list of the major packages in the documentation: https://metcalcpy.readthedocs.io/en/latest/Users_Guide/installation.html.  The version of Python we used for development is 3.6.3, to "match" what NCO had installed on their HPC clusters.  There are a lot of extraneous third-party packages and in the beta1 release for 2.0, we will have identified the minimal number of third-party packages.  

For the METplotpy requirements, we don't have a requirements file yet, but here is a list of the major packages we use in our Github actions to run all the plot tests:  

pip install numpy==1.19.5
        pandas==1.1.5
        plotly==4.9.0
        imutils==0.5.3
        kaleido==0.2.1
        matplotlib==3.3.4
        netcdf4==1.5.7
        opencv-python==4.5.5.62
        pytest==6.2.5
        PyYAML==5.4.1
        scikit-image==0.16.2
        scikit-learn==0.23.2
        scipy==1.5.3
        statsmodels==0.12.1
        xarray==0.16


Here is a more extensive list of the major third-party packages and their versions:
https://metplotpy.readthedocs.io/en/latest/Users_Guide/installation.html

We will have a requirements.txt file available for METplotpy for the beta1 release.

**Reply by cpapadop1** [reply]
Awesome, thanks, I will try to get this working with the METcalcpy setup later this week, and look forward to the next release.

**Comment by bikegeek** [comment]
I created a Github issue in the METplotpy repository: 
https://github.com/dtcenter/METplotpy/issues/223

to find a solution for the Python 3.8 version of the code.



Can you please close out this discussion?   We can communicate through the "comments" section of the Github issue.
